### DISCUSSION TITLE
Strategy for handling large cdk stacks that exceed CloudFormation Template size limit of 1000000 bytes

### DISCUSSION THREAD

**Discussion by Shogan** [aws/aws-cdk#22529] [General]
I'm getting the following error in CloudFormation.
Error [ValidationError]: Template may not exceed 1000000 bytes in size.

We're using aws-cdk to manage custom 'environments' for a project. The environments are already split up into nested stacks for resources, but in some cases there can be a large number of these nested stack resources.

The resulting template therefore gets many (e.g. 100x) AWS::CloudFormation::Stack resources (the nested stack resources).

To the point where it looks like the resulting overall stack template for the CloudFormation changeset exceeeds 1000000 bytes and fails validation as it is not allowed (the limit is 1MB for templates in S3).

Does anyone have a mitigation strategy to avoid this? E.g. I've heard of preprocessors that can minify the JSON changesets, but have not seen a way of integrating this with aws-cdk.

**Comment by Shogan** [comment]
I have been experimenting with some techniques and found a relatively simple way to achieve what I'm after, using out of the box features of cdk. Basically:

- `cdk synth` into a target output_directory
- run a simple nodejs script that iterates over templates in the output directory and runs a minify on the JSON to remove whitespace
- `cdk deploy -a output_directory`

In a large sized template (just shy of 1MB in size) I was able to reduce it down to 750KB using this initial, simple technique.

Full npm script as an example:

```
"npm run build && cdk synth --json --output cdk-output && node minify-cfn-templates.js && cdk deploy -a cdk-output"
```

**Reply by pahud** [reply]
@Shogan This is interesting! Thank you for the sharing!

**Reply by stowns** [reply]
Any idea if minifying templates is something that CDK plans to support natively? The CDK metadata greatly increases the file size while still remaining under the 500 Cfn resource

**Reply by cslogan-red** [reply]
We're using the CDK to manage several stacks to automate our infrastructure provisioning and I'm barely hitting this limit trying to deploy our REST API stack (the REST API is the only thing in this stack); CDK should provide a minification flag or minify output templates by default.

**Reply by JoshMcCullough** [reply]
Rather than minifying the template, CDK should simply "not care" about the template size. Or, at least not unless it is very large. 1MB is not that large!

**Comment by HansFalkenberg-Visma** [comment]
There now seems to be `suppressTemplateIndentation`.

https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.StackProps.html

Setting it got me down from 1.03 MB to 827901 bytes.

If that leaves you just above the limit, you could use minification as suggested above to remove newlines too.

**Reply by JoshMcCullough** [reply]
This helps, but having to do this is still pretty silly. A 1MB stack is not that large, really.

**Comment by taabrcr** [comment]
Hi guys, thanks a lot for the tips and workarounds. +1 on requesting AWS  to provide a flag for minification or doing this by deafult.

**Reply by khushail** [reply]
@taabrcr , you could open a [feature request](https://github.com/aws/aws-cdk/issues/new?assignees=&labels=feature-request%2Cneeds-triage&projects=&template=feature-request.yml&title=%28module+name%29%3A+%28short+issue+description%29) with CDK and help prioritizing it with upvotes from the community.

**Reply by JoshMcCullough** [reply]
There is a flag, [see above](https://github.com/aws/aws-cdk/discussions/22529#discussioncomment-9536836).

**Comment by brizaldi** [comment]
Is there any plan to increase the size limit? We're currently using CDK, and this limit is severely affecting our workflow.
We have already applied the minify & `suppressTemplateIndentation` option as suggested above, but this is only a temporary solution, as the template size will continue to grow along with our resources.

**Reply by JoshMcCullough** [reply]
I ended up splitting my stack into about 5 smaller stacks. This seems to be the way to go.

**Comment by chanh-t** [comment]
There's a property for this. If you set `suppressTemplateIndentation` to `true` for the stackProperty, it'll reduce the whitespaces.

There's a thread for this discussion as well as the link to the commit as well.
Thread:
[suppressTemplateIndentation](https://github.com/aws/aws-cdk/issues/18694)
Cr:
https://github.com/aws/aws-cdk/commit/b70595686e0742691bf64ce80bd18ea26694400d

**Comment by Erry91** [comment]
Not sure if anyone found any other cool trick regarding this. For example we encountered this problem when we decided to keep the managed login branding style (the template and manifests) tracked in IaC in our CDK project.
The main problem here is that the generated cdk template easily exceeds 1MB even with the indentation suppression flag.
